feat(lib): methods for measuring time and duration

alexzurbonsen opened the pull request:
Committed without pre-commit hook because tests are still missing.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Enhancement (project structure, spelling, grammar, formatting)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.


### A description of the changes proposed in the Pull Request
I am opening this PR because I want to figure out if I am heading in the right direction.

So far, the plugins I have seen in if are basically functions or operators that map variables to other variables. They are not measuring anything but perform a calculation.

But if you actually want to measure, you are also interested in the timestamp and the duration of the measurement. And you don't want to write that by hand into your manifest files. Thus, I wrote this timer methods that can manipulate timestamp and duration.

Does that conform with how you imagine if should be used? If so, do you think this plugin should find its way into the unofficial plugins? I would have to add tests then.

<!-- Make sure tests and lint pass on CI. -->

manushak commented on src/lib/timer/start.ts:
```diff
@@ -0,0 +1,54 @@
+  const validateInput = (input: PluginParams) => {
+    const schema = z
+      .object({
+        resets: z.array(z.boolean()),
```
instead of using `refine`, you can write like so
` resets: z.array(z.boolean()).min(1, {message: ERROR_MESSAGE_RESETS}),`
in this case, the error message needs to update

alexzurbonsen commented on src/lib/timer/README.md:
```diff
@@ -0,0 +1,66 @@
+      path: '@alexzurbonsen/if-webpage-models-cjs'
+    "measure-webpage":
+      method: MeasureWebpage
+      path: '@alexzurbonsen/if-webpage-models-cjs'
```
@manushak Thanks for your comments. I will address them as soon as possible. 

About the installation: How did you try it? I didn't publish the plugin on npm. But with `npm install -g git+https://github.com/alexzurbonsen/if-webpage-models-cjs.git` it should work.

And yes, the `MeasureWebpage` plugin is not in the repo. It is still work in progress. But happy to contribute it once it is ready. But it should already work, if you want to test the manifest file (given that you replace `models` with `plugins` 😅 )

alexzurbonsen commented:
I addressed your comments, but still need to add some tests.

alexzurbonsen commented:
Hi @manushak, a short update on the changes I pushed: 

a) I added tests.
b) I changed the approach: `TimerStart` is now just writting a `timer/start` value to each input but does not change `timestamp` or `duration`. This logic moved to `TimerStop`. That has the advantage that `TimerStop` can check if `TimerStart` was run before by looking for the `timer/start` field. Of course, that also changes the input and return values of both methods. `TimerStop` consumes the resets now.
c) The example manifest from the README now uses a simple plugin that can execute shell commands (instead of the measurement plugin). I opened a PR for that plugin here #59.

jmcook1186 commented:
Hi @alexzurbonsen i'm struggling to understand the intended use case for this - if you want to measure the time it takes to execute a pipeline, why not do this externally (i.e. time IF execution)? If you have a need for this, then you are of course encouraged to maintain the plugins in your own repository, but I'm not sure it's something we want to maintain in the IF repo.

alexzurbonsen commented:
Hi @jmcook1186,
The idea is, that once you actually measure something with a plugin, and not just put measured values as inputs into the pipeline, then at least I would want to automatically write timestamp and measurement duration in the inputs. In this [draft PR](the need for this can be seen in this draft PR: https://github.com/Green-Software-Foundation/if-unofficial-plugins/pull/63) the measure-webpage plugin is loading a webpage and measuring its page weight for example. I would want to know at which time that measurement was done.

But as  I also wrote in the summary of the PR, I am not sure you had imagined IF to be used in this way, because plugins I have seen are pretty static so far and not measuring things.

jmcook1186 commented:
Similar to the other PR - this seems like a potentially useful idea, but it's not really something we need in our standard library - recommend maintaining this in an external repo. We'll be creating a public registry for external plugins soon.
